File: go/dsa/string/StringUtil.go
package string

import (
	"strings"
)

// StringUtil é uma estrutura para agrupar as funções de utilidade de string
type StringUtil struct{}

// IsSubsequence verifica se 'seq' é uma subsequência de 'str'
func IsSubsequence(str, seq string) bool {
	i, j := 0, 0 // i percorre 'str' e j percorre 'seq'

	// Enquanto não chegamos ao fim de 'str' ou 'seq'
	for i < len(str) && j < len(seq) {
		// Se o caractere atual de 'str' é igual ao de 'seq', avançamos em 'seq'
		if str[i] == seq[j] {
			j++
		}
		// Avançamos sempre em 'str'
		i++
	}
	// Se j atingiu o comprimento de 'seq', então 'seq' é uma subsequência de 'str'
	return j == len(seq)
}

// RemoveVowels remove todas as vogais de uma string
func RemoveVowels(str string) string {
	// Define um conjunto de caracteres que são considerados vogais
	vowels := "aeiou"
	var result strings.Builder // strings.Builder para construir a nova string

	// Percorre cada caractere na string
	for _, ch := range str {
		// Se o caractere atual não é uma vogal, adiciona-o ao resultado
		if !strings.ContainsRune(vowels, ch) {
			result.WriteRune(ch)
		}
	}
	// Retorna a string sem vogais
	return result.String()
}

// FirstNonRepeatingCharacter retorna o índice do primeiro caractere não repetido em 'str'
func FirstNonRepeatingCharacter(str string) int {
	// Mapa para armazenar a frequência de cada caractere
	characterFrequencyMap := make(map[rune]int)
	chars := []rune(str) // Converte a string para um slice de runas

	// Conta a frequência de cada caractere
	for _, ch := range chars {
		characterFrequencyMap[ch]++
	}

	// Percorre os caracteres novamente para encontrar o primeiro não repetido
	for i, ch := range chars {
		if characterFrequencyMap[ch] == 1 {
			return i // Retorna o índice do primeiro caractere não repetido
		}
	}
	// Se nenhum caractere com frequência 1 foi encontrado, retorna -1
	return -1
}

File: csharp/src/dsa/string/StringUtil.cs
namespace dsa.str;

public class StringUtil
{
    // static void Main()
    // {
    //     Console.WriteLine(FirstNonRepeatingCharacter("abafbabe"));
    //     Console.WriteLine(isSubsequence("abcde", "ace"));
    //     Console.WriteLine(removeVowels("hello world"));
    // }

    public static bool isSubsequence(String str, String seq)
    {
        int i = 0; // Índice para percorrer a string 'str'
        int j = 0; // Índice para percorrer a string 'seq'

        // Enquanto não chegamos ao fim de 'str' ou 'seq'
        while (i < str.Length && j < seq.Length)
        {
            // Se o caractere atual em 'str' é igual ao caractere atual em 'seq', incrementamos 'j'
            if (str[i] == seq[j])
                j++;

            // Sempre incrementamos 'i' para avançar para o próximo caractere em 'str'
            i++;
        }
        // Se 'j' é igual ao comprimento de 'seq', então todos os caracteres de 'seq' foram encontrados em 'str' na ordem correta
        return j == seq.Length;
    }

    public static String removeVowels(String str)
    {
        // Define um conjunto de caracteres que são considerados vogais
        HashSet<char> vowels = ['a', 'e', 'i', 'o', 'u'];

        // Cria uma lista para armazenar os caracteres da string sem as vogais
        List<char> chars = [];

        // Percorre cada caractere na string
        foreach (char ch in str)
        {
            // Se o caractere atual não é uma vogal, adiciona-o à lista
            if (!vowels.Contains(ch))
                chars.Add(ch);
        }
        // Converte a lista de caracteres de volta para uma string e retorna
        return new string(chars.ToArray());
    }

    public static int FirstNonRepeatingCharacter(string str)
    {
        // Cria um dicionário para armazenar a frequência de cada caractere na string
        Dictionary<char, int> characterFrequencyMap = [];
        // Converte a string para um array de caracteres
        char[] chars = str.ToCharArray();

        // Percorre cada caractere no array
        foreach (char ch in chars)
        {
            // Se o caractere já está no dicionário, incrementa sua frequência
            if (characterFrequencyMap.ContainsKey(ch))
                characterFrequencyMap[ch]++;
            // Se o caractere não está no dicionário, adiciona-o com frequência 1
            else
                characterFrequencyMap[ch] = 1;
        }

        // Percorre cada caractere no array novamente
        for (int i = 0; i < chars.Length; i++)
        {
            // Se a frequência do caractere atual é 1, retorna seu índice
            if (characterFrequencyMap[chars[i]] == 1)
                return i;
        }
        // Se nenhum caractere com frequência 1 foi encontrado, retorna -1
        return -1;
    }
}

